processing/src/main/java/org/apache/druid/query/aggregation/SingleValueAggregatorFactory.java

```
@@ -0,0 +1,206 @@
+ * {@link SingleValueBufferAggregator} to work with different ColumnTypes
+ */
+@JsonTypeName("singleValue")
+public class SingleValueAggregatorFactory extends AggregatorFactory
+{
+
+  @JsonProperty
+  @JsonInclude(JsonInclude.Include.NON_NULL)
+  private final String name;
+
+  @JsonProperty
+  @JsonInclude(JsonInclude.Include.NON_NULL)
+  private final String fieldName;
+
+  @JsonProperty
+  @JsonInclude(JsonInclude.Include.NON_NULL)
+  private final ColumnType columnType;
+
+  public static final int DEFAULT_MAX_STRING_SIZE = 1024;
+
+  @JsonCreator
+  public SingleValueAggregatorFactory(
+      @JsonProperty("name") String name,
+      @JsonProperty("fieldName") final String fieldName,
+      @JsonProperty("columnType") final ColumnType columnType
+  )
+  {
+    Preconditions.checkNotNull(name, "Must have a valid, non-null aggregator name");
+    Preconditions.checkNotNull(fieldName, "Must have a valid, non-null fieldName");
+
+    this.name = name;
+    this.fieldName = fieldName;
+    this.columnType = columnType;
+  }
+
+  @Override
+  public Aggregator factorize(ColumnSelectorFactory metricFactory)
+  {
+    ColumnValueSelector selector = metricFactory.makeColumnValueSelector(fieldName);
+    return new SingleValueAggregator(selector);
+  }
+
+  @Override
+  public BufferAggregator factorizeBuffered(ColumnSelectorFactory metricFactory)
+  {
+    ColumnValueSelector selector = metricFactory.makeColumnValueSelector(fieldName);
+    ColumnCapabilities columnCapabilities = metricFactory.getColumnCapabilities(fieldName);
+    ColumnType columnType = new ColumnType(columnCapabilities.getType(), null, null);
+    return new SingleValueBufferAggregator(selector, columnType);
+  }
+
+  @Override
+  public Comparator getComparator()
+  {
+    throw DruidException.defensive("Single Value Aggregator would not have more than one row to compare");
+  }
+
+  @Override
+  @Nullable
+  public Object combine(@Nullable Object lhs, @Nullable Object rhs)
```

What would happen if `lhs` represents something empty (i.e. hasn't encountered a single row), while `rhs` is something that has encountered a single row. In that case `rhs` should be returned. Similarly, the cases for when `lhs` represents something which means a single row, while `rhs` is empty; or when both are empty). I am not sure how we would encode "that something is empty", hence the confusion.

This also is at odds with `getCombiningFactory`, since that doesn't throw upon being called, and returns a correct implementation. I assumed that both should have the same semantics. The calling patterns might be such that its okay to not implement the .combine() and throw an exception here, and if so, I think there should be some comment explaining as to why.

---

IMO this would never be called as there's nothing to group. This runs on the broker level on the result of the subquery. We should throw an exception in the `getCombiningFactory` as well because that also is also not needed

---

@somu-imply Shouldn't we throw the same exception on the `getCombiningAggregator` as well then?

---

Also, going through the code path, wouldn't combine be called if the subquery is run on two segments, where one segment returns a single result (matching some hypothetical filter in the subquery), while the second segment doesn't have any row matching the result, and therefore returns `null`?

---

Laksh, I tried creating a small datasource on multiple segments. combine method was never invoked.
getCombiningAggregator was invoked and was used to get the correct column name

---

Thanks! I think we should add a comment then, stating the combine is never called, and the combining aggregator is called for the fieldName only.